File: backend/GameApp.Application/Services/ItemServices/ItemUpdateService.cs
using GameApp.Application.UseCases.ItemUseCases;
using GameApp.Domain.Entities.Items;
using GameApp.Domain.Repositories;

namespace GameApp.Application.Services.ItemServices;

public class ItemUpdateService : ItemUpdateUseCase
{
    private readonly IItemRepository _repo;

    public ItemUpdateService(IItemRepository repo) => _repo = repo;


    public async Task<Item?> UpdateItem(Guid id, Item item)
    {
        var existingItem = await _repo.FetchByIdAsync(id);
        if (existingItem is null)
        {
            Console.WriteLine($"Item to update with id {id} not found.");
            return null;
        }

        // If the item name is going to change
        if (!existingItem.GetName().Equals(item.GetName()))
        {
            var itemWithSameName = await _repo.FetchByName(item.GetName());
            if (itemWithSameName is not null)  // item name must be unique in the collection
            {
                Console.WriteLine($"Item with name '{item.GetName()}' already exists. Cannot update item");
                return null;
            }
        }

        return await _repo.UpdateAsync(id, item);
    }
}

File: frontend/gameapp-frontend/src/application/services/item-services/item-update-service.ts
import type { IItemUpdateUseCase } from "../../usecases/item-use-cases/item-update-use-case";
import type { IItemRepository } from "../../repositories/iitem-repository";
import { Item } from "../../../domain/entities/items/item";

export class ItemUpdateService implements IItemUpdateUseCase {
    constructor(private itemRepository: IItemRepository) { }

    async updateItem(id: string, item: Item): Promise<Item> {
        const updatedItem = await this.itemRepository.update(id, item);
        return updatedItem;
    }
}
